### PR TITLE
Ensure resource profile changes are applied correctly without any issues

### PR DESCRIPTION
**Pull request by malayparida2000**
```
• Determine resource profile change by comparing sc.Spec.ResourceProfile
  with sc.Status.LastAppliedResourceProfile. If they are different, then
  a resource profile change is in progress.
• Mark the ContinueUpgradeAfterChecksEvenIfNotHealthy flag on the
  cephCluster CR to false if a resource profile change is in progress.
• During the below process mark the storageCluster phase as Progressing.
  The storageCluster conditions will show "New resource profile is being
  applied".
• Keep checking if any of the pods with the new resource profile
  label are in pending state for more than 2 minutes. If yes, then
  assume that the resource profile change has failed. Mark the 
  storageCluster phase as Error and the storageCluster conditions will 
  show "New resource profile failed to apply, please revert to the last 
  working resource profile".
• Start Checking if the ceph dameons(mgr,mon,osd,mds,rgw) have the
  correct number of running pods where each pod has the current
  resource profile label.
• If the above checks are successful, the new resource profile has
  been applied successfully. Set the resource profile in the status &
  Mark the storageCluster phase to Ready.
```

### CODE REVIEW

**Comment by malayparida2000** [comment]
The first 3 commit s are of this https://github.com/red-hat-storage/ocs-operator/pull/2242 PR.
Don't merge before that PR.
/hold

**Comment by malayparida2000** [comment]
Removing hold as the mentioned PR is now merged.
/unhold

File: controllers/storagecluster/cephcluster.go
Comment (travisn) [inline controllers/storagecluster/cephcluster.go:465]: The `sc.Status.ResourceProfile` is only updated after the successful reconcile of the new resource profile, right? So this is comparing the new profile with the previous profile (if it changed)?

Comment (travisn) [reply controllers/storagecluster/cephcluster.go:465]: Will this cause all upgraded clusters (the first time upgrading to 4.15) to set this to `false`? Let's keep this as `true` if we're applying the default profile and if the applied profile is blank.

File: controllers/storagecluster/resourceprofile.go
Comment (travisn) [inline controllers/storagecluster/resourceprofile.go:68]: Instead of verifying the OSD pod count with the label, how about just checking if there are any pods that don't have the expected resource profile label? Or a similar question, is why are the OSD pods different from checking the mon or mgr pods?

File: controllers/storagecluster/cephcluster.go
Comment (malayparida2000) [reply controllers/storagecluster/cephcluster.go:465]: Yeah, the thought was that if the values on the status & the spec are not the same that means we are in the process of applying the resource profile that's in the spec.

Comment (malayparida2000) [reply controllers/storagecluster/cephcluster.go:465]: No, when someone upgrades to 4.15, They don't have the resource profile either in their spec or status. So two blank values are equal, and the flag would be true only.

File: controllers/storagecluster/resourceprofile.go
Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go]: Hm, Initially I thought about verifying them each storage device set wise, but seems like it's not really required & just adding complexity.

**Review by travisn — COMMENTED** [review]
/lgtm

File: controllers/storagecluster/resourceprofile.go
Comment (umangachapagain) [inline controllers/storagecluster/resourceprofile.go:17]: How is this value determined? 20 minutes is a lot of time for pods to restart with new resource requirements. Can we decrease it?

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/resourceprofile.go
Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go:17]: @travisn How much time do you think will be sufficient for a typical cluster, for rook to process all the daemon restarts?

File: controllers/storagecluster/reconcile.go
Comment (malayparida2000) [reply controllers/storagecluster/reconcile.go:481]: In the condition, the message will read "New resource profile is being applied",
So the reason can be something different. As we have reasons related to the word reconcile only https://github.com/red-hat-storage/ocs-operator/blob/1f682da09475c6419ff9f104fecd0caa7197b46d/api/v1/storagecluster_types.go#L559-L561

Comment (umangachapagain) [reply controllers/storagecluster/reconcile.go:481]: Those only occur at specific conditions. But, reconcile happens every few seconds.
So we are almost always in "ReconcileOngoing" state, but we are only sometime in "ReconcileFailed" or "ApplyingResourceProfileChanges" state.
Even during cluster expansion, we set condition to "progressing" with message that "cluster is explanding", but we do not define "ReconcileOngoing" as a reason.

Comment (malayparida2000) [reply controllers/storagecluster/reconcile.go:481]: Hmm, Then let's put the reason as Reconcile failed, As we are actually waiting for the desired state & it hasn't reached there in this reconcile, and we have to requeue and check again.

File: controllers/storagecluster/resourceprofile.go
Comment (travisn) [reply controllers/storagecluster/resourceprofile.go:17]: Timeouts are a challenge because it depends on a number of factors, such as whether images are being pulled, how big the cluster is (number of OSDs), etc. In this case of updating the resources at least we should only be updating existing daemons and not deploying new daemons where images might need to be pulled. What about a more dynamic set of checks such as these?
1. If any daemons are in pending state, consider it failed immediately since it's likely pending due to lack of resources.
2. If no daemons have been updated in some short timeout, then consider it failed. For example, keep track of the number of daemons with the new profile label. If it hasn't increased for two minutes, then consider it failed. 

The first condition should actually be sufficient to catch the cause, so we may not need any timeout checking from the second bullet.

Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go]: Have removed the timeout, And implemented another logic to check if any of the coming pods with the resource profile label are stuck in pending for more than 2 mins. PTAL
@travisn @umangachapagain

File: controllers/storagecluster/resourceprofile.go
Comment (travisn) [inline controllers/storagecluster/resourceprofile.go:34]: If the pod has been pending for less than two minutes, is there a retry again to wait and see if it gets out of the pending state? We don't want to continue if any pod is pending, even if less than two minutes. The two minutes was just suggested as a short timeout to consider the reconcile failed instead of waiting and retrying.

Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go:34]: Ok, Adding a logic to return here itself if it's pending for less than 2 mins.

Comment (malayparida2000) [reply controllers/storagecluster/resourceprofile.go]: I meant if less than 2 min, Return & retry. If more than 2 mins just declare failed.

**Review by travisn — COMMENTED** [review]
/lgtm assuming the testing goes well to confirm the expected behavior.

**Comment by malayparida2000** [comment]
> /lgtm assuming the testing goes well to confirm the expected behavior.

Testing wise, I have covered profile changing in both up and down directions,
And intentionally setting too high a requirement in performance mode & changing to it, to see the behavior.
Also the recovery from this situation to a previous profile.
Hope this covers the main use cases.

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**
